webui/src/Views/ParameterFormView.elm

```diff
@@ -416,73 +454,112 @@ newParameterValueView template parameterInfos maybeInstanceParameterForm enabled
-                                    )
-                                ]
-                                [ icon "glyphicon glyphicon-copy" [] ]
+                                []
                             ]
-                         else
-                            []
+                            (if (isSecret) then
+                                [ a
+                                    [ class "input-group-addon"
+                                    , attribute "role" "button"
+                                    , onClick (ToggleNewInstanceSecretVisibility template.id parameter)
+                                    , id <| String.concat [ "new-instance-form-parameter-secret-visibility-", template.id, "-", parameter ]
+                                    ]
+                                    [ icon
+                                        (String.concat
+                                            [ "glyphicon glyphicon-eye-"
+                                            , (if secretVisible then
+                                                "close"
+                                               else
+                                                "open"
+                                              )
+                                            ]
+                                        )
+                                        []
+                                    ]
+                                , a
+                                    [ class "input-group-addon"
+                                    , attribute "role" "button"
+                                    , attribute
+                                        "onclick"
+                                        (String.concat
+                                            [ "copy('"
+                                            , parameterValue
+                                            , "')"
+                                            ]
+                                        )
+                                    ]
+                                    [ icon "glyphicon glyphicon-copy" [] ]
+                                ]
+                             else
+                                []
+                            )
                         )
-                    )
-                ]
+                    ]
+                    (case maybeErrMsg of
+                        Nothing -> []
+                        Just msg ->
+                            [span
+                                [class "help-block"]
+                                [text msg]
+                            ]
+                     )
+                 )
+
+getErrorMsg: String -> DataType -> Maybe String
+getErrorMsg value dataType =
+    case dataType of
+        -- we do not validate string and raw fields
+        StringParam -> Nothing
+        RawParam -> Nothing
+        NumericParam ->
+            let
+                expr = regex "^-?[0-9]+(\\.[0-9]+)?$"
```

Comment:
Also it sounds like a bad idea to check the param again. I think we should store the parameters with correct types and then simply decode them.

As Scala doesn't like Maps with multiple types we can either

- store the type information somewhere and have dynamic casting (not so optimal), or
- have one map per parameter type which is type safe and insert them in a multi-branch when injecting the template configuration. I.e. we have a `rawParameters` map, a `stringParameters` map, and a `numericParameters` map? To avoid too many values in the case class we can always have a wrapper case class grouping all the parameter values / infos. Not sure if this one is trivial to implement, also because we map the template.conf structure 1 to 1 to our case class, but I prefer something type safe and sound over storing strings and validating / parsing them on the fly. This gets especially tricky when you think about other parameter types like sets, etc. How are you going to represent them safely as strings and parse? The logic will be very error prone.

Reply:
Ah I see there is the need to also encode the parameters in the front-end. Sorry for my confusion. Then you actually need two validation steps, or? But the back-end should rely on JSON serialization and the front-end should send appropriate types over, i.e. a number as a number and not quoted as a string.

You can use toInt and toFloat http://package.elm-lang.org/packages/elm-lang/core/latest/String

Reply:
We also discussed this. It seems a bit involved right now to send stuff as specific types instead of strings (because of the Map stuff that you said). Also, JSON may not be enough for some future types we may need to support for example positive integers (for CPU/memory etc.) or ranged integers (for ports) and so on. I would suggest we take that up as a separate issue maybe.

Reply:
It seems involved but storing everything as strings in a huge map and then trying to convert it with custom logic seems wrong. It might look nice and "clean" for numbers but even simple structures like lists require you to duplicate basic JSON/HOCON serialization logic. Also we should fail fast when loading the templates and not at the point where we try to use them.

I also think that the argument for positive integers is not a very strong one. Custom types holding positive integers are usually a wrapper for normal integers and not strings. IMO we should make use of the existing types as much as possible and add additional stuff on top as required.

Here's what I think could be a good solution:

Take

```
parameters = {
  "id" = {
    type = string
  }
  "json" = {
    type = raw
  }
  "cpu" = {
    type = number
  }
}
```

and parse it to [`TemplateConfig.TemplateInfo`](https://github.com/FRosner/cluster-broccoli/blob/master/server/src/main/scala/de/frosner/broccoli/templates/TemplateConfig.scala) as it is already done. In https://github.com/FRosner/cluster-broccoli/blob/master/server/src/main/scala/de/frosner/broccoli/models/ParameterInfo.scala#L20 we add the logic to build up the individual properly typed parameters, also validating the default parameters to fail fast. Having one map type for each parameter type sounds like additional work, but it's a simple solution IMO.

The front-end code then needs to send the appropriate types. If we have something like a positive int, we specify custom JSON encoders and decoders.

@Gerrrr what do you think?

Reply:
> Having one map type for each parameter type sounds like additional work, but it's a simple solution IMO. However, it will blow up when we add more and more parameter types so maybe there is a smarter solution. We can have Map[String, (Any, ParameterType)] and then cast it according to the type before using it.

I am somehow averse to casts (also will probably need to do it repeatedly everywhere we need it). I would rather move towards an inheritance chain of some sort maybe. But I can't visualize this very precisely right now.

Reply:
Inheritance chain sounds scary :D

Reply:
>As Scala doesn't like Maps with multiple types we can either

We could define an enum with supported types and say that the map's values are from this enum. Every case from the enum would have its sanitization logic.

Reply:
That's what I was suggesting. But the compiler wants a type for the map. And this would be any then. Sounds like a use case for union types to me but I don't know if we want to go down that road... http://milessabin.com/blog/2011/06/09/scala-union-types-curry-howard/ :D

Reply:
> As Scala doesn't like Maps with multiple types we can either

Or have a trait for a template parameter type and classes extending this trait. Then, the map would be of type e.g. `[String, TemplateParameterType]`.

Reply:
You can always use pattern matching to match the types.

```scala
val parameterValues: Map[String, Any] = ???
parameterValues.mapValues {
  case s: String => ???
  case n: BigDecimal => ???
  case _ => ???
}
```

If we want to limit the type scope, having a generic parameter value container like it's done with `JsValue` is another way.

Something like `ParameterValue` as the super type and then subtypes like `StringParameterValue`, `RawParameterValue`. Then pattern matching is much nicer.

Reply:
> Or have a trait for a template parameter type and classes extending this trait. Then, the map would be of type e.g. [String, TemplateParameterType].

It can't be a map to the type. It needs to be a map to the value. But otherwise I agree.

Reply:
So it would be `Map[String, ParameterValue]` and then we have different `ParameterValue` implementations which provide JSON/HOCON serdes accordingly and this gives us validation out of the box.

Reply:
> Or have a trait for a template parameter type and classes extending this trait. Then, the map would be of type e.g. [String, TemplateParameterType].

This was what I was talking about exactly (maybe I used the wrong word but it is still inheritance isn't it? :D)

> It can't be a map to the type. It needs to be a map to the value. But otherwise I agree. The map to the type is given by the parameter infos.

And yes this. In the end, we end up with a type map and a value map. I was thinking maybe we could combine this (since you have the types anyway with the trait stuff). This is the part I said I couldn't visualise.

Reply:
I think it doesn't make sense to combine it from a conceptual point of view as the parameter infos are available after template parsing but the values can be changed during operation time by user requests.

I agree with having two maps as discussed above. Ping me if you want to discuss another WIP or final version.

I think we don't need a call now.